### PR TITLE
feature(dspy): dspy-integration-with-langfuse

### PR DESCRIPTION
**Pull request by xucailiang**
As mentioned in issue [#1179](https://github.com/stanfordnlp/dspy/issues/1179) , I wrote some code to integrate langfuse.
 when configured correctly, It will support the display of data from OpenAI, AzureOpenAI and Ollama on the Langfuse dashboard

### CODE REVIEW

File: dsp/modules/gpt3.py
Comment (denisergashbaev) [inline dsp/modules/gpt3.py:9]: what do you mean that they have made it compatible for us?

File: dsp/modules/ollama.py
Comment (denisergashbaev) [inline dsp/modules/ollama.py:10]: i believe we should not have a tight coupling to langfuse in DSPy. It should be some sort of a hadler that is passed to DSPy: similar to how langfuse integrates with LangChain

File: dsp/modules/ollama.py
Comment (denisergashbaev) [inline dsp/modules/ollama.py:146]: this is hardcoded for "Ollama" and also tracing is hardcoded. Would it not be possible to do tracing outside of DSPy code?

**Review by denisergashbaev — COMMENTED** [review]
Hey @xucailiang I have left some comments. I am learning DSPy too and just wondering whether the proposed integration could be done in a more elegant way

**Comment by xucailiang** [comment]
> 嘿@xucailiang我留下了一些评论。我也在学习 DSPy，只是想知道是否可以以更优雅的方式完成所提议的集成

**Comment by xucailiang** [comment]
> Hey @xucailiang I have left some comments. I am learning DSPy too and just wondering whether the proposed integration could be done in a more elegant way

I have rewritten the code, the new calling method is more elegant, I realize that all LLM reasoning can be collected to langfuse, but it is difficult for me to test all the code. I have already tested my submission, I don't know if I can get some help

**Comment by denisergashbaev** [comment]
@xucailiang hey! I will check it out this weekend

**Comment by xucailiang** [comment]
> @xucailiang hey! I will check it out this weekend

Thanks for your time, I'm not sure if we'll wrap up this pull request over the weekend. If you have any suggestions or ideas to improve this, please feel free to share them with me at your earliest convenience. I'd really appreciate it!

**Comment by xucailiang** [comment]
> @xucailiang hey! I will check it out this weekend

hi，I wonder if this discussion is still ongoing?

**Comment by canada4663** [comment]
+1, also interested in this!

**Comment by denisergashbaev** [comment]
> > @xucailiang hey! I will check it out this weekend
> 
> hi，I wonder if this discussion is still ongoing?

Hi! Unfortunately I have not had time to run it locally yet. I will do it. However, I am not a DSPy maintainer, just another user, so all my comments are just my comments :) I wonder whether it would make sense to ping one of the contributers so that they can take a look as well

**Comment by xucailiang** [comment]
@arnavsinghvi11 hi,I was wondering if you could review this PR so I can continue integrating langfuse on other modules.Thank you.

**Comment by arnavsinghvi11** [comment]
Hi @xucailiang , thanks for opening this PR! Exciting to see the Langfuse integration with DSPy. 

The PR looks on track so far, but I agree with your point that it's difficult to add this for all LM providers. I would suggest exploring ways to wrap the Langfuse logic not to each provider class but rather with the base [lm.py class ](https://github.com/stanfordnlp/dspy/blob/main/dsp/modules/lm.py) so any user using any LM provider can declare the langfuse tracker as an arg and then have it popped before the LM request is made. 

Let me know if that makes sense and if you have any additional questions. Feel free to mention me here again when the PR is ready to review as well!

**Comment by xucailiang** [comment]
> Hi @xucailiang , thanks for opening this PR! Exciting to see the Langfuse integration with DSPy.
> 
> The PR looks on track so far, but I agree with your point that it's difficult to add this for all LM providers. I would suggest exploring ways to wrap the Langfuse logic not to each provider class but rather with the base [lm.py class ](https://github.com/stanfordnlp/dspy/blob/main/dsp/modules/lm.py) so any user using any LM provider can declare the langfuse tracker as an arg and then have it popped before the LM request is made.
> 
> Let me know if that makes sense and if you have any additional questions. Feel free to mention me here again when the PR is ready to review as well!

Indeed, I have modified some of the code for this, I am glad you pointed out this oversight.

**Comment by xucailiang** [comment]
hi,I noticed that dspy has been updated recently. Will the integration of langfuse appear in the next update? @arnavsinghvi11

**Comment by arnavsinghvi11** [comment]
hi @xucailiang , i think the latest changes look good. the only thing i'd call out is currently this integration is only supported for openai, azure-openai and ollama. would it be possible to keep the `self.tracker = tracker` within the [LM](https://github.com/stanfordnlp/dspy/blob/5aca3f23b632be57883168b1867aee852fc027ca/dsp/modules/lm.py) class, and the tracker decorator on the [base LM call method](https://github.com/stanfordnlp/dspy/blob/5aca3f23b632be57883168b1867aee852fc027ca/dsp/modules/lm.py#L135)?  

This is what I meant by keeping it generalized to all LM providers, and if it is not a necessary arg in the LM request payload (which it likely won't be for any provider), it can be popped before the request is made. 

lmk if that makes sense and when the PR is ready to merge!

**Comment by xucailiang** [comment]
> hi @xucailiang , i think the latest changes look good. the only thing i'd call out is currently this integration is only supported for openai, azure-openai and ollama. would it be possible to keep the `self.tracker = tracker` within the [LM](https://github.com/stanfordnlp/dspy/blob/5aca3f23b632be57883168b1867aee852fc027ca/dsp/modules/lm.py) class, and the tracker decorator on the [base LM call method](https://github.com/stanfordnlp/dspy/blob/5aca3f23b632be57883168b1867aee852fc027ca/dsp/modules/lm.py#L135)?
> 
> This is what I meant by keeping it generalized to all LM providers, and if it is not a necessary arg in the LM request payload (which it likely won't be for any provider), it can be popped before the request is made.
> 
> lmk if that makes sense and when the PR is ready to merge!
> 
> (there are also a couple of merge conflicts)

Of course, `self.tracker = tracker` is already in base lm, I also added sample documentation to let users know that they can manually call the tracker when using any module.

> hi @xucailiang , i think the latest changes look good. the only thing i'd call out is currently this integration is only supported for openai, azure-openai and ollama. would it be possible to keep the `self.tracker = tracker` within the [LM](https://github.com/stanfordnlp/dspy/blob/5aca3f23b632be57883168b1867aee852fc027ca/dsp/modules/lm.py) class, and the tracker decorator on the [base LM call method](https://github.com/stanfordnlp/dspy/blob/5aca3f23b632be57883168b1867aee852fc027ca/dsp/modules/lm.py#L135)?
> 
> This is what I meant by keeping it generalized to all LM providers, and if it is not a necessary arg in the LM request payload (which it likely won't be for any provider), it can be popped before the request is made.
> 
> lmk if that makes sense and when the PR is ready to merge!
> 
> (there are also a couple of merge conflicts)

Great advice，I re-optimized the code and made the following changes:

1. Keep `self.tracker` in base **LM** and retain the tracker decorator to facilitate new module integration
2. Add a method to call tracker in base **LM**.all modules except openai and azure-openai can also call the tracker
3. supplementary document

File: README.md
Comment (arnavsinghvi11) [inline README.md:83]: can we actually add this to the DSPy docs instead of in the README? Feel free to add this to https://github.com/stanfordnlp/dspy/tree/main/docs/api/language_model_clients

File: dsp/modules/ollama.py
Comment (arnavsinghvi11) [inline dsp/modules/ollama.py:61]: are these lines needed if Ollama already inherits from LM?

File: dsp/trackers/tracker_decorator.py
Comment (arnavsinghvi11) [inline dsp/trackers/tracker_decorator.py:2]: maybe im just missing it but where is the tracker_decorator used?

**Review by arnavsinghvi11 — CHANGES_REQUESTED** [review]

**Comment by arnavsinghvi11** [comment]
Thanks for the additions @xucailiang ! left some minor comments for finishing touches before its ready to merge. mainly, let's add the documentation and examples for Langfuse in its separate doc, not in the README. lgtm after changes!

**Comment by xucailiang** [comment]
> Thanks for the additions @xucailiang ! left some minor comments for finishing touches before its ready to merge. mainly, let's add the documentation and examples for Langfuse in its separate doc, not in the README. lgtm after changes!

Yes, I have made some changes according to your suggestions. Please review it again. Thank you for your valuable time.

**Comment by arnavsinghvi11** [comment]
made a couple edits to the docs. Merged now! Thanks @xucailiang this integration!

File: dsp/modules/azure_openai.py
Comment (bettlebrox) [inline dsp/modules/azure_openai.py:13]: I believe this causes dspy to use the langfuse openai wrapper if langfuse is simply installed, even if env variables are not configured. 

Considering langfuse also have a [decorator integration method](https://langfuse.com/docs/sdk/python/decorators) I think the wrapper should only be used if langfuse is configured via env variables or offer some other flag to switch of the use of the langfuse openai wrapper

**Pull request merged**
